File: DSA/done/2466 - countGoodStrings.py
# // https://leetcode.com/problems/count-ways-to-build-good-strings/

# Memo
class Solution:
    def countGoodStrings(self, low: int, high: int, zero: int, one: int) -> int:
        dp = {}
        MOD = 10**9 + 7 

        def dfs(i):
            if i > high:
                return 0
            if i in dp:
                return dp[i]

            res = 0
            if i >= low:
                res += 1

            res += dfs(i + zero) 
            res %= MOD
            res += dfs(i + one)
            res %= MOD
            dp[i] = res
            return res

        return dfs(0) % MOD
    
# Dp
class Solution:
    def countGoodStrings(self, low: int, high: int, zero: int, one: int) -> int:
        dp = [0] * (high + max(zero, one) + 1)
        MOD = 10**9 + 7 

        for i in range(high, -1, -1):
            dp[i] = 0 if i < low else 1
            dp[i] += dp[i + zero] 
            dp[i] %= MOD
            dp[i] += dp[i + one]
            dp[i] %= MOD

        return dp[0] % MOD

File: DSA/done/2466 - countGoodStrings.cpp
// https://leetcode.com/problems/count-ways-to-build-good-strings/

#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

// Memo
class Solution {
public:
    const int MOD = 1'000'000'007;
    unordered_map<int, int> memo;

    int dfs(int i, int low, int high, int zero, int one) {
        if (i > high) return 0;
        if (memo.find(i) != memo.end()) return memo[i];

        int res = (i >= low) ? 1 : 0;
        res += dfs(i + zero, low, high, zero, one);
        res %= MOD;
        res += dfs(i + one, low, high, zero, one);
        res %= MOD;

        memo[i] = res;
        return res;
    }

    int countGoodStrings(int low, int high, int zero, int one) {
        return dfs(0, low, high, zero, one);
    }
};

// Dp
class Solution {
public:
    int countGoodStrings(int low, int high, int zero, int one) {
        const int MOD = 1'000'000'007;
        int maxSize = high + max(zero, one) + 1;
        vector<int> dp(maxSize, 0);

        for (int i = high; i >= 0; i--) {
            dp[i] = (i < low) ? 0 : 1;
            if (i + zero < maxSize) {
                dp[i] += dp[i + zero];
                dp[i] %= MOD;
            }
            if (i + one < maxSize) {
                dp[i] += dp[i + one];
                dp[i] %= MOD;
            }
        }

        return dp[0];
    }
};